File: generated_code/optimized/balanced/cpp/4.cpp
#include <vector>
#include <cmath>
#include <cstddef>
using namespace std;

float mean_absolute_deviation(vector<float> numbers) {
    const size_t n = numbers.size();
    if (n == 0) return 0.0f;

    const float* p = numbers.data();
    double sum = 0.0;
    for (size_t i = 0; i < n; ++i) sum += p[i];

    const float mean = static_cast<float>(sum / n);

    double dev_sum = 0.0;
    for (size_t i = 0; i < n; ++i) dev_sum += std::fabs(p[i] - mean);

    return static_cast<float>(dev_sum / n);
}

File: generated_code/optimized/balanced/java/4.java
import java.util.List;
import java.util.RandomAccess;

class Solution {
    public double meanAbsoluteDeviation(List<Double> numbers) {
        int n = numbers.size();
        if (n == 0) return Double.NaN;

        double sum = 0.0;
        if (numbers instanceof RandomAccess) {
            for (int i = 0; i < n; i++) {
                sum += numbers.get(i);
            }
            double mean = sum / n;
            double absSum = 0.0;
            for (int i = 0; i < n; i++) {
                absSum += Math.abs(numbers.get(i) - mean);
            }
            return absSum / n;
        } else {
            for (Double num : numbers) {
                sum += num;
            }
            double mean = sum / n;
            double absSum = 0.0;
            for (Double num : numbers) {
                absSum += Math.abs(num - mean);
            }
            return absSum / n;
        }
    }
}
